File: rust/src/patch_radiation.rs
// patch_radiation.rs
// Rust implementation of patch radiation calculations, ported from the original Python source.
// This file is intended to match the structure and intent of the Python version in pysrc/umepr/patch_radiation.py.

use std::f32::consts::PI;

// Vectorized functions removed as they were unused.


pub fn longwave_from_sky_pixel(
    lsky_side: f32,
    lsky_down: f32,
    patch_azimuth: f32,
) -> (f32, f32, f32, f32, f32, f32) {
    let deg2rad = PI / 180.0;
    let ldown_sky = lsky_down;
    let lside_sky = lsky_side;

    let least = if !(180.0..=360.0).contains(&patch_azimuth) {
        lside_sky * ((90.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lsouth = if patch_azimuth > 90.0 && patch_azimuth < 270.0 {
        lside_sky * ((180.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lwest = if patch_azimuth > 180.0 && patch_azimuth < 360.0 {
        lside_sky * ((270.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lnorth = if !(90.0..=270.0).contains(&patch_azimuth) {
        lside_sky * ((0.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    (lside_sky, ldown_sky, least, lsouth, lwest, lnorth)
}

// Vectorized functions removed as they were unused.


pub fn longwave_from_veg_pixel(
    steradian: f32,
    angle_of_incidence: f32,
    angle_of_incidence_h: f32,
    patch_altitude: f32,
    patch_azimuth: f32,
    ewall: f32,
    ta: f32,
) -> (f32, f32, f32, f32, f32, f32) {
    let sbc = 5.67051e-8;
    let deg2rad = PI / 180.0;
    let vegetation_surface = (ewall * sbc * (ta + 273.15).powi(4)) / PI;
    let cos_alt = (patch_altitude * deg2rad).cos();

    let lside_veg = vegetation_surface * steradian * angle_of_incidence;
    let ldown_veg = vegetation_surface * steradian * angle_of_incidence_h;

    let least = if !(180.0..=360.0).contains(&patch_azimuth) {
        vegetation_surface * steradian * cos_alt * ((90.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lsouth = if patch_azimuth > 90.0 && patch_azimuth < 270.0 {
        vegetation_surface * steradian * cos_alt * ((180.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lwest = if patch_azimuth > 180.0 && patch_azimuth < 360.0 {
        vegetation_surface * steradian * cos_alt * ((270.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lnorth = if !(90.0..=270.0).contains(&patch_azimuth) {
        vegetation_surface * steradian * cos_alt * ((0.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    (lside_veg, ldown_veg, least, lsouth, lwest, lnorth)
}

// Vectorized functions removed as they were unused.


pub fn longwave_from_buildings_pixel(
    steradian: f32,
    angle_of_incidence: f32,
    angle_of_incidence_h: f32,
    patch_azimuth: f32,
    sunlit_patch: bool,
    shaded_patch: bool,
    azimuth_difference: f32,
    solar_altitude: f32,
    ewall: f32,
    ta: f32,
    tgwall: f32,
) -> (f32, f32, f32, f32, f32, f32, f32, f32) {
    let sbc = 5.67051e-8;
    let deg2rad = PI / 180.0;
    let sunlit_surface = (ewall * sbc * (ta + tgwall + 273.15).powi(4)) / PI;
    let shaded_surface = (ewall * sbc * (ta + 273.15).powi(4)) / PI;

    let (lside_sun, lside_sh, ldown_sun, ldown_sh, mask_val) =
        if (azimuth_difference > 90.0) && (azimuth_difference < 270.0) && (solar_altitude > 0.0) {
            let sunlit = if sunlit_patch { 1.0 } else { 0.0 };
            let shaded = if shaded_patch { 1.0 } else { 0.0 };
            let lside_sun = sunlit * sunlit_surface * steradian * angle_of_incidence;
            let lside_sh = shaded * shaded_surface * steradian * angle_of_incidence;
            let ldown_sun = sunlit * sunlit_surface * steradian * angle_of_incidence_h;
            let ldown_sh = shaded * shaded_surface * steradian * angle_of_incidence_h;
            let mask_val = sunlit * sunlit_surface + shaded * shaded_surface;
            (lside_sun, lside_sh, ldown_sun, ldown_sh, mask_val)
        } else {
            let lside_sun = 0.0;
            let lside_sh = shaded_surface * steradian * angle_of_incidence;
            let ldown_sun = 0.0;
            let ldown_sh = shaded_surface * steradian * angle_of_incidence_h;
            let mask_val = shaded_surface;
            (lside_sun, lside_sh, ldown_sun, ldown_sh, mask_val)
        };

    let least = if !(180.0..=360.0).contains(&patch_azimuth) {
        mask_val * steradian * angle_of_incidence * ((90.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lsouth = if patch_azimuth > 90.0 && patch_azimuth < 270.0 {
        mask_val * steradian * angle_of_incidence * ((180.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lwest = if patch_azimuth > 180.0 && patch_azimuth < 360.0 {
        mask_val * steradian * angle_of_incidence * ((270.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lnorth = if !(90.0..=270.0).contains(&patch_azimuth) {
        mask_val * steradian * angle_of_incidence * ((0.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };

    (
        lside_sun, lside_sh, ldown_sun, ldown_sh, least, lsouth, lwest, lnorth,
    )
}

// Vectorized functions removed as they were unused.


pub fn longwave_from_buildings_wall_scheme_pixel(
    voxel_table: ndarray::ArrayView2<f32>,
    voxel_map_val: usize,
    steradian: f32,
    angle_of_incidence: f32,
    angle_of_incidence_h: f32,
    patch_azimuth: f32,
) -> (f32, f32, f32, f32, f32, f32, f32, f32) {
    let deg2rad = PI / 180.0;

    let patch_radiation = if voxel_map_val > 0 {
        // This is a simplification. In a real scenario, you'd have a more robust
        // way to map voxel_map_val to the correct row in voxel_table.
        // Assuming voxel_map_val is a 1-based index corresponding to the row.
        voxel_table.row(voxel_map_val - 1)[1]
    } else {
        0.0
    };

    let lside = patch_radiation * steradian * angle_of_incidence;
    let ldown = patch_radiation * steradian * angle_of_incidence_h;
    let lside_sh = 0.0;
    let ldown_sh = 0.0;

    let least = if !(180.0..=360.0).contains(&patch_azimuth) {
        patch_radiation * steradian * angle_of_incidence * ((90.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lsouth = if patch_azimuth > 90.0 && patch_azimuth < 270.0 {
        patch_radiation * steradian * angle_of_incidence * ((180.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lwest = if patch_azimuth > 180.0 && patch_azimuth < 360.0 {
        patch_radiation * steradian * angle_of_incidence * ((270.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lnorth = if !(90.0..=270.0).contains(&patch_azimuth) {
        patch_radiation * steradian * angle_of_incidence * ((0.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };

    (
        lside, lside_sh, ldown, ldown_sh, least, lsouth, lwest, lnorth,
    )
}

// Vectorized functions removed as they were unused.


pub fn reflected_longwave_pixel(
    steradian: f32,
    angle_of_incidence: f32,
    angle_of_incidence_h: f32,
    patch_azimuth: f32,
    ldown_sky: f32,
    lup: f32,
    ewall: f32,
) -> (f32, f32, f32, f32, f32, f32) {
    let deg2rad = PI / 180.0;
    let reflected_radiation = (ldown_sky + lup) * (1.0 - ewall) * 0.5 / PI;
    let lside_ref = reflected_radiation * steradian * angle_of_incidence;
    let ldown_ref = reflected_radiation * steradian * angle_of_incidence_h;
    let least = if !(180.0..=360.0).contains(&patch_azimuth) {
        reflected_radiation
            * steradian
            * angle_of_incidence
            * ((90.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lsouth = if patch_azimuth > 90.0 && patch_azimuth < 270.0 {
        reflected_radiation
            * steradian
            * angle_of_incidence
            * ((180.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lwest = if patch_azimuth > 180.0 && patch_azimuth < 360.0 {
        reflected_radiation
            * steradian
            * angle_of_incidence
            * ((270.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };
    let lnorth = if !(90.0..=270.0).contains(&patch_azimuth) {
        reflected_radiation
            * steradian
            * angle_of_incidence
            * ((0.0 - patch_azimuth) * deg2rad).cos()
    } else {
        0.0
    };

    (lside_ref, ldown_ref, least, lsouth, lwest, lnorth)
}

File: pysrc/solweig/physics/patch_radiation.py
"""
Patch-level radiation helpers — **reference implementation only**.

Not called by the production ``calculate()`` API. The fused Rust pipeline
computes patch radiation internally.

Retained for readability, tests, and validation against UMEP.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from ..constants import KELVIN_OFFSET, SBC

_DEG2RAD = np.pi / 180


def _cardinal_components(
    base_radiation: NDArray[np.floating], patch_azimuth: float
) -> tuple[NDArray[np.float32], NDArray[np.float32], NDArray[np.float32], NDArray[np.float32]]:
    """Split radiation into E/S/W/N components based on patch azimuth.

    The azimuth boundaries determine which cardinal hemisphere a patch
    contributes to, weighted by the cosine of the angular offset.

    Returns:
        Tuple of (Least, Lsouth, Lwest, Lnorth) arrays.
    """
    shape = np.shape(base_radiation)
    Least = np.zeros(shape, dtype=np.float32)
    Lsouth = np.zeros(shape, dtype=np.float32)
    Lwest = np.zeros(shape, dtype=np.float32)
    Lnorth = np.zeros(shape, dtype=np.float32)

    if patch_azimuth < 180:
        Least = base_radiation * np.cos((90 - patch_azimuth) * _DEG2RAD)
    if (patch_azimuth > 90) and (patch_azimuth < 270):
        Lsouth = base_radiation * np.cos((180 - patch_azimuth) * _DEG2RAD)
    if (patch_azimuth > 180) and (patch_azimuth < 360):
        Lwest = base_radiation * np.cos((270 - patch_azimuth) * _DEG2RAD)
    if (patch_azimuth > 270) or (patch_azimuth < 90):
        Lnorth = base_radiation * np.cos((0 - patch_azimuth) * _DEG2RAD)

    return Least, Lsouth, Lwest, Lnorth


def shortwave_from_sky(sky, angle_of_incidence, lumChi, steradian, patch_azimuth, cyl):
    """Calculates the amount of diffuse shortwave radiation from the sky for a patch with:
    angle of incidence = angle_of_incidence
    luminance = lumChi
    steradian = steradian"""

    # Diffuse vertical radiation
    diffuse_shortwave_radiation = sky * lumChi * angle_of_incidence * steradian

    return diffuse_shortwave_radiation


def longwave_from_sky(sky, Lsky_side, Lsky_down, patch_azimuth):
    Ldown_sky = sky * Lsky_down
    Lside_sky = sky * Lsky_side
    Least, Lsouth, Lwest, Lnorth = _cardinal_components(sky * Lsky_side, patch_azimuth)
    return Lside_sky, Ldown_sky, Least, Lsouth, Lwest, Lnorth


def longwave_from_veg(
    vegetation, steradian, angle_of_incidence, angle_of_incidence_h, patch_altitude, patch_azimuth, ewall, Ta
):
    """Longwave radiation from a vegetation patch."""
    vegetation_surface = (ewall * SBC * ((Ta + KELVIN_OFFSET) ** 4)) / np.pi
    Lside_veg = vegetation_surface * steradian * angle_of_incidence * vegetation
    Ldown_veg = vegetation_surface * steradian * angle_of_incidence_h * vegetation

    base = vegetation_surface * steradian * np.cos(patch_altitude * _DEG2RAD) * vegetation
    Least, Lsouth, Lwest, Lnorth = _cardinal_components(base, patch_azimuth)

    return Lside_veg, Ldown_veg, Least, Lsouth, Lwest, Lnorth


def longwave_from_buildings(
    building,
    steradian,
    angle_of_incidence,
    angle_of_incidence_h,
    patch_azimuth,
    sunlit_patches,
    shaded_patches,
    azimuth_difference,
    solar_altitude,
    ewall,
    Ta,
    Tgwall,
):
    sunlit_surface = (ewall * SBC * ((Ta + Tgwall + KELVIN_OFFSET) ** 4)) / np.pi
    shaded_surface = (ewall * SBC * ((Ta + KELVIN_OFFSET) ** 4)) / np.pi

    if (azimuth_difference > 90) and (azimuth_difference < 270) and (solar_altitude > 0):
        Lside_sun = sunlit_surface * sunlit_patches * steradian * angle_of_incidence * building
        Lside_sh = shaded_surface * shaded_patches * steradian * angle_of_incidence * building
        Ldown_sun = sunlit_surface * sunlit_patches * steradian * angle_of_incidence_h * building
        Ldown_sh = shaded_surface * shaded_patches * steradian * angle_of_incidence_h * building

        # Cardinal components: sum of sunlit + shaded contributions
        base_sun = sunlit_surface * sunlit_patches * steradian * angle_of_incidence * building
        base_sh = shaded_surface * shaded_patches * steradian * angle_of_incidence * building
        Le_sun, Ls_sun, Lw_sun, Ln_sun = _cardinal_components(base_sun, patch_azimuth)
        Le_sh, Ls_sh, Lw_sh, Ln_sh = _cardinal_components(base_sh, patch_azimuth)
        Least = Le_sun + Le_sh
        Lsouth = Ls_sun + Ls_sh
        Lwest = Lw_sun + Lw_sh
        Lnorth = Ln_sun + Ln_sh
    else:
        Lside_sh = shaded_surface * steradian * angle_of_incidence * building
        Lside_sun = np.zeros_like(Lside_sh)
        Ldown_sh = shaded_surface * steradian * angle_of_incidence_h * building
        Ldown_sun = np.zeros_like(Lside_sh)

        base = shaded_surface * steradian * angle_of_incidence * building
        Least, Lsouth, Lwest, Lnorth = _cardinal_components(base, patch_azimuth)

    return Lside_sun, Lside_sh, Ldown_sun, Ldown_sh, Least, Lsouth, Lwest, Lnorth


def longwave_from_buildings_wallScheme(
    voxelMaps, voxelTable, steradian, angle_of_incidence, angle_of_incidence_h, patch_azimuth
):
    unique_ids = list(np.unique(voxelMaps)[1:])
    lw_rad_dict = dict(voxelTable.loc[unique_ids, "LongwaveRadiation"])
    patch_radiation = np.vectorize(lw_rad_dict.get)(voxelMaps).astype(float)
    patch_radiation[np.isnan(patch_radiation)] = 0

    Lside = patch_radiation * steradian * angle_of_incidence
    Ldown = patch_radiation * steradian * angle_of_incidence_h

    base = patch_radiation * steradian * angle_of_incidence
    Least, Lsouth, Lwest, Lnorth = _cardinal_components(base, patch_azimuth)

    Lside_sh = np.zeros_like(Lside)
    Ldown_sh = np.zeros_like(Ldown)
    return Lside, Lside_sh, Ldown, Ldown_sh, Least, Lsouth, Lwest, Lnorth


def reflected_longwave(
    reflecting_surface, steradian, angle_of_incidence, angle_of_incidence_h, patch_azimuth, Ldown_sky, Lup, ewall
):
    reflected_radiation = ((Ldown_sky + Lup) * (1 - ewall) * 0.5) / np.pi
    Lside_ref = reflected_radiation * steradian * angle_of_incidence * reflecting_surface
    Ldown_ref = reflected_radiation * steradian * angle_of_incidence_h * reflecting_surface

    base = reflected_radiation * steradian * angle_of_incidence * reflecting_surface
    Least, Lsouth, Lwest, Lnorth = _cardinal_components(base, patch_azimuth)

    return Lside_ref, Ldown_ref, Least, Lsouth, Lwest, Lnorth


def patch_steradians(L_patches):
    """'This function calculates the steradians of the patches"""

    # Degrees to radians
    deg2rad = np.pi / 180

    # Unique altitudes for patches
    skyalt, skyalt_c = np.unique(L_patches[:, 0], return_counts=True)

    # Altitudes of the Robinson & Stone patches
    patch_altitude = L_patches[:, 0]

    # Calculation of steradian for each patch
    # Build scalar lookup once to avoid array->scalar coercion warnings.
    count_by_altitude = {float(alt): float(count) for alt, count in zip(skyalt, skyalt_c, strict=False)}
    steradian = np.zeros((patch_altitude.shape[0]), dtype=np.float32)
    for i in range(patch_altitude.shape[0]):
        band_count = count_by_altitude[float(patch_altitude[i])]
        # If there are more than one patch in a band
        if band_count > 1:
            steradian[i] = ((360 / band_count) * deg2rad) * (
                np.sin((patch_altitude[i] + patch_altitude[0]) * deg2rad)
                - np.sin((patch_altitude[i] - patch_altitude[0]) * deg2rad)
            )
        # If there is only one patch in band, i.e. 90 degrees
        else:
            steradian[i] = ((360 / band_count) * deg2rad) * (
                np.sin((patch_altitude[i]) * deg2rad) - np.sin((patch_altitude[i - 1] + patch_altitude[0]) * deg2rad)
            )

    return steradian, skyalt, patch_altitude
